packages/jest-worker/src/Farm.ts

```diff
@@ -78,80 +81,82 @@ export default class Farm {
     });
   }
 
-  private _getNextJob(workerId: number): QueueChildMessage | null {
+  private _getNextTask(workerId: number): QueueChildMessage | null {
     let queueHead = this._queue[workerId];
 
-    while (queueHead && queueHead.request[1]) {
+    while (queueHead && queueHead.task.request[1]) {
       queueHead = queueHead.next || null;
     }
 
     this._queue[workerId] = queueHead;
 
-    return queueHead;
+    return queueHead && queueHead.task;
   }
 
   private _process(workerId: number): Farm {
-    if (this.isLocked(workerId)) {
+    if (this._isLocked(workerId)) {
       return this;
     }
 
-    const job = this._getNextJob(workerId);
+    const task = this._getNextTask(workerId);
 
-    if (!job) {
+    if (!task) {
       return this;
     }
 
     const onEnd = (error: Error | null, result: unknown) => {
-      job.onEnd(error, result);
-      this.unlock(workerId);
+      task.onEnd(error, result);
+
+      this._unlock(workerId);
       this._process(workerId);
     };
 
-    this.lock(workerId);
+    task.request[1] = true;
 
-    this._callback(workerId, job.request, job.onStart, onEnd);
-
-    job.request[1] = true;
+    this._lock(workerId);
+    this._callback(workerId, task.request, task.onStart, onEnd);
 
     return this;
   }
 
   private _enqueue(task: QueueChildMessage, workerId: number): Farm {
+    const item = {task, next: null};
```

Comment:
The core of the fix is only this line; where instead of pushing to all queues the `task` object, we create one specifically for each queue, meaning that when doing `.next` in the tail of a linked list, we don't affect all others).

The rest of the PR are "cosmetic" changes, and the test.